helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -235,8 +376,94 @@ PartitionInfo get(String stateUnitKey) {
     }
   }
 
+  class ResourceGroupInfo {
+    // store PartitionInfo for each partition for all resources in the resource group
+    HashMap<String, PartitionInfo> partitionInfoMap;
+    // stores the Set of Instances in a given state for all resources in the resource group
+    HashMap<String, Set<InstanceConfig>> stateInfoMap;
+    // stores the ResourceInfo for each resource with different instance tag.
+    HashMap<String, ResourceInfo> tagResourceInfoMap;
+
+    public ResourceGroupInfo() {
+      partitionInfoMap = new HashMap<String, RoutingTableProvider.PartitionInfo>();
+      stateInfoMap = new HashMap<String, Set<InstanceConfig>>();
+      tagResourceInfoMap = new HashMap<String, ResourceInfo>();
+    }
+
+    public void addEntry(String resourceTag, String stateUnitKey, String state, InstanceConfig config) {
+      // add
+      if (!stateInfoMap.containsKey(state)) {
+        Comparator<InstanceConfig> comparator = new Comparator<InstanceConfig>() {
```

Comment:
Declare the comparator outside since it seems to be identical to the one in `ResourceInfo`